web/components/elements/DropdownSearch.tsx

```diff
@@ -0,0 +1,95 @@
+import { FunctionComponent } from "preact";
+import { useCallback, useEffect, useState } from "preact/hooks";
+import style from "./DropdownSearch.module.scss";
+
+interface Props<T> {
+    searchFn: (searchTerm: string, additionalProps: T) => Promise<DropItem[]>;
+    searchProps: T,
+    placeholder?: string,
+    onChange: (value: string|null) => void;
+    onError?: (error: unknown) => void;
+}
+
+export interface DropItem {
+    value: string;
+    title: string;
+    imageSrc?: string;
+    description?: string;
+}
+
+const DEBOUNCE_TIMEOUT_MS = 300;
+
+export const DropdownItem: FunctionComponent<DropItem&{onPicked?: (value: string) => void}> = ({ imageSrc, title, value, description, onPicked }) => {
+    // Need placeholder image.
+    return <li className={`card ${style.dropdownItem} ${imageSrc ? style.hasImg : ''}`} role="button" onClick={() => onPicked?.(value)}>
+            { imageSrc && <img className={style.itemImage} src={imageSrc} /> }
+            <div>
+                <p className={style.title}>{title}</p>
+                <p className={style.value}>{value}</p>
+                <p className={style.description}>{description}</p>
+            </div>
+    </li>;
+};
+
+export const DropdownSearch = function<T>({searchFn, searchProps, onChange, onError, placeholder}: Props<T>) {
+    const [debounceTimer, setDebounceTimer] = useState<number|undefined>(undefined);
+    const [selectedItem, setSelectedItem] = useState<DropItem|null>();
+    const [searchTerm, setSearchTerm] = useState<string>("");
+    const [results, setResults] = useState<DropItem[]>([]);
+
+    // Reset if the search properties are altered.
+    useEffect(() => {
+        setSearchTerm("");
+        setSelectedItem(null);
+    }, [searchProps]);
+
+    // Search whenever the term is updated.
+    useEffect(() => {
+        if (searchTerm.trim().length === 0) {
+            return;
+        }
+        if (selectedItem) {
+            // Clear any selected items
+            setSelectedItem(null);
+            onChange(null);
+        }
+
+        // Browser types
+        setDebounceTimer(setTimeout(() => {
+            console.log("searching for ", searchTerm);
+            searchFn(searchTerm, searchProps).then(result => {
+                setResults(result);
+            }).catch(err => {
+                onError?.(err);
+            })
```

Comment:
Minor nitpick but likely a rare edge that users know to deal with: Once the search has been started, a slow network fetch will override a fast one.
It will also set the results if `onItemPicked` has set the results to `[]` or `searchProps` changed in the meantime.